src/compiler/checker.ts

```diff
@@ -10305,11 +10400,17 @@ namespace ts {
 
         function getTypeAtPosition(signature: Signature, pos: number): Type {
             return signature.hasRestParameter ?
-                pos < signature.parameters.length - 1 ? getTypeOfSymbol(signature.parameters[pos]) : getRestTypeOfSignature(signature) :
-                pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length - 1) :
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length, anyType);
         }
 
         function assignContextualParameterTypes(signature: Signature, context: Signature, mapper: TypeMapper) {
+            if (context.thisType) {
+                if (signature.declaration.kind !== SyntaxKind.ArrowFunction) {
+                    // do not contextually type thisType for ArrowFunction. 
+                    signature.thisType = context.thisType;
+                }
+            }

```

Comment:
`--strictThis` needs to be on too, or the interface needs to be `interface I { a: number, m(this: this): number }`.

I haven't tried running the language service in an actual editor yet. Let me try that and see whether the explicit version works..